### ISSUE TITLE
[✨] Allow binding `this` to the store in useStore

### ISSUE THREAD

**Issue by wmertens** [BuilderIO/qwik#5087]
### Is your feature request related to a problem?

It would be nice to have have functions in stores be bound to the store object.

### Describe the solution you'd like

I think a `bind` option to useStore would be nice. It could even be `true` by default because the majority use case for functions in the store is using the store object itself.

This can be used orthogonally with `deep` and `reactive`.

```tsx
import { component$, useStore, $, useTask$ } from '@builder.io/qwik';

export default component$(() => {
  const store = useStore({
    count: 1,
    inc: $(function () {console.log('I have a this:', !!this);this.count++})
  }, {bind: true})

  return <p>Count: {store.count} <button onClick$={store.inc}>one up</button></p>;
});
```

### Describe alternatives you've considered

letting the deserializer do it, but then the function is only bound when it's deserialized.

### Additional context

This allows making zustand-like APIs.

**Comment by justoverclockl** [comment]
pr opened, hope it helps:

https://github.com/BuilderIO/qwik/pull/5088

**Comment by mhevery** [comment]
I assume this workaround is too wordy

```typescript
import { component$, useStore, $, useTask$ } from '@builder.io/qwik';

export default component$(() => {
  const store = useStore({
    count: 1,
    inc: null!
  })
  store.inc = $(function () {store.count++})

  return <p>Count: {store.count} <button onClick$={store.inc}>one up</button></p>;
});
```

Wondering if something like this would work:

```typescript
import { component$, useStore, $, useTask$ } from '@builder.io/qwik';

export default component$(() => {
  const store = useStore({
    count: 1,
    inc: storeMethod$(function () {store.count++})
  })

  return <p>Count: {store.count} <button onClick$={store.inc}>one up</button></p>;
});
```

The semantics of `method$` would be such that it would ensure that `store.inc` would pre-bind it to the `store`.

**Comment by wmertens** [comment]
I found how to do it!

https://qwik.builder.io/playground/#v=1.1.5&buildMode=production&f=Q0o0xobG2BKNjgIQwWpdilMQJDjS8vOBoVENbhlYgYuuWqAcUFQP0liwhVuoAdIE0gYs9q0UDHXAvMy8ZCsFkKkgQ%2BG69MCKtLVrNYGKwATEsmqwuA5IVy3QZLh6LpQE7QyxAaK4FhiY4FYOonKvBmm3A%2FoHWFnC20CjaX8op30A

```tsx
import { component$, $, useStore } from '@builder.io/qwik';

export default component$(() => {
  const foo = {store: null}
  foo.store = useStore({
    count: 1,
    inc: $(()=> {foo.store.count++})
  })
  const {count, inc} = foo.store

  return <p>Count: {count} <button onClick$={inc}>one up</button></p>;
});
```

Would the optimizer be able to take
```tsx
const store = useStore({ count: 0, inc: $(() => store.inc++) })
```
and convert it to
```tsx
const __ref = {}
ref.store = useStore({ count: 0, inc: $(() => __ref.store.inc++) })
const store = __ref.store
```
?

**Comment by mhevery** [comment]
We don't want to teach optimizers such tricks, as it would lead to weird corner cases. How should the optimizer know when it is safe to convert `this` to `__ref`? I can imagine lot's of ways to break it, which will result in bugs being filed. The simpler, the better.

**Comment by wmertens** [comment]
@mhevery this would only apply when it's encountering a self-reference during conversion. `this` is not used.

```js
const store = useStore({ count: 0, inc: $(() => store.inc++) })
```

is valid JS

but the conversion does

```js
const store = useStore({ count: 0, inc: qrl(() => ..., [store]) })
```

which is not valid due to reference before initialization. Replacing `const` with `var` makes it valid JS but the QRL function will be getting `[undefined]` so it won't help.

By making a temporary object as a stable ref, the conversion becomes

```js
const __ref = {};
__ref.store = useStore({
    count: 1,
    inc: /*#__PURE__*/ qrl(()=>import("./s_wp3wbze0ge0.js"), "s_wP3wBzE0Ge0", [
        __ref
    ])
});
const store = __ref.store
```

which is correct JS and the serializer is happy too.

**Comment by wmertens** [comment]
The optimizer would be applying a transform that makes it accept valid JS code.

**Comment by wmertens** [comment]
@mhevery I understand and share your caution, but I think that in this particular case it's fine, because the binding isn't happening invisibly, it's part of the scope. The second example you give won't be bound by anything, not in regular JS and not after optimizing.

I think that this is a corner case where the optimizer encounters valid JS and due to its regular transform, it breaks the JS.

By adding a temporary ref when needed (or even every time), the optimizer generates code that is equivalent to what was there before. The temporary ref is only visible to the optimizer and the QRL encapsulation.

I believe this allows writing useful JS in an intuitive way.

**Comment by wmertens** [comment]
In fact, this same technique allows you to define recursive QRL functions:

```tsx
import { component$, $ } from '@builder.io/qwik';

const fibModuleScope = n => n > 1 ? (fibModuleScope(n-1)+fibModuleScope(n-2)) : 1

export default component$(() => {
  const __ref = {} as any
  __ref.fib$ = $(async n => n > 1 ? ((await __ref.fib$(n-1))+(await __ref.fib$(n-2))) : 1)
  const {fib$} = __ref

  return <p>Fib 5 {fibModuleScope(5)}<br/> Fib$ 5 {fib$(5)} </p>;
});
```

https://qwik.builder.io/playground/#v=1.1.5&f=Q0o0xuaW2BKNjgL%2BdAPxYFpmkm9%2BSmlOKrgOA3o1D%2BTTPAU7BUMFewUNVGmNPF1DTW0MMSNNTQUrBUPikiLE2vh4YIEHtK26FljGgCo0oAxYTA9ougpQQkUDUvagOkcDUk0gVIJdpKmNTRzoKrCzNOGWVoNkaoGGg9VxoeQLt8wkBVOwCmS%2FmWrW2iQV6dspAKVVoPIqIFFgXIzmnqGbewA

**Comment by mhevery** [comment]
Workaround: 

Sorry, here is a working example: https://stackblitz.com/edit/qwik-starter-hyo58s?file=README.md,src%2Froutes%2Findex.tsx

```typescript
import { component$, $, useStore } from '@builder.io/qwik';

interface State {
  count: number;
  increment(): Promise<void>;
}

export default component$(() => {
  const state = useStore<State>({
    count: 0,
    increment: $(function (this: State) {
      this.count++;
    }),
  });
  return (
    <button onClick$={() => state.increment()}>Count: {state.count}</button>
  );
});
```

**Comment by wmertens** [comment]
@mhevery is that workaround better than using an extra object for containing the scope?

**Comment by brandonpittman** [comment]
Functions at the top level of the store object that reference `this` work fine but I'd like to have a function inside of a nested object in the store. Is there a way to reference state at the top level of the store? `this` becomes the nested object instead of the root.

Is this just a JS issue?

**Comment by wmertens** [comment]
@brandonpittman you can do it with this technique https://github.com/BuilderIO/qwik/issues/5087#issuecomment-1707185010

I really do think this is safe to implement in the optimizer but the fact that it's in Rust makes that hard.

**Comment by brandonpittman** [comment]
@wmertens 

> you can do it with this technique https://github.com/BuilderIO/qwik/issues/5087#issuecomment-1707185010

You're correct. This seems to work fine. Is there something I'm missing that makes it unsafe? The docs only make mention of the `this` style of function definition.

**Comment by wmertens** [comment]
No nothing I can think of.

**Comment by ritbl** [comment]
If we split state and actions, we don't need `this` trick. Sandbox [here](https://qwik.builder.io/examples/introduction/hello-world/#f=7VZNSwNBDL37K0IpdBYGD55k2S4sxYMiFe2C5xULVaR7KIiyzn83ecnMDlVKr4X2snQmk4%2BX5CVZ0VxfpUQhLQM9Pt0fqJiYhqnPisWnAX2o1saxtGSagjhQEdIqbVUQuP8T1C7eQZ4HkpyX4mnlmPUxjqIGgP3ZvzFCrEoTEb17Zr6LmuZUrdqmvWGSYxZ43VH%2F8s5c7KlZtLcPy9XeeS1J4zrSNz9k%2FYG%2Fhee7riSHpcdkcG26%2FEUqAdt%2BEPx83GvwpHZSb6lMBkh56tTjkjo1UHAGObRjWlet7SkKmeEMkSqB7f%2FgjugJKuWnGZvd9ZvtTGInCvoxB8W8i8IpXVM3NmXMmnaiCcrbS4wwbFUQsNtQqBn5MEZZL1kbMTtpmFAQhJ%2F0XBfXcV8bFB3DgjkL3rlJ87H%2BWn9PilAvNt2WB43oSXsv6%2BIGUifOvHeqvPcL)


```

const useStoreWithActions = <STATE extends object, ACTIONS extends object>(
  s: STATE | (() => STATE),
  a: (store: STATE) => ACTIONS,
) => {
  const store = useStore<STATE>(s);
  return {store, actions: a(store)};
}

interface NameStore {
  name: string;
}

interface NameStoreActions {
  setName: QRL<(newValue: string) => void>
}

export default component$(() => {
  const {store, actions} = useStoreWithActions<NameStore, NameStoreActions>(
    {
      name: 'John',
    },
    (store) => ({
      setName: $(function (newValue) {
        store.name = newValue
      })
    })
  );

  return (<>
    <p>{store.name}</p>
    <button onClick$={() => actions.setName("Alexey")}>Change name</button>
  </>)
```

**Comment by wmertens** [comment]
@ritbl that's still a truck and it has more moving parts...

**Comment by ritbl** [comment]
@wmertens what is the best practice for state mutation? Is it expected to mutate state directly or use these kinds of setters?

**Comment by wmertens** [comment]
Here's a nice example of how to do self-contained objects in serialized state. I still think that the optimizer could make this automatically.

https://github.com/QwikDev/qwik/issues/6139#issuecomment-2067930487
